### ISSUE TITLE
Proposal for future copy / view semantics in indexing operations

### ISSUE THREAD

**Issue by TomAugspurger** [pandas-dev/pandas#36195]
Update: concrete proposal for Copy-on-Write: https://docs.google.com/document/d/1ZCQ9mx3LBMy-nhwRl33_jgcvWo9IWdEfxDNQ2thyTb0/edit?usp=sharing

---
[_original post_]

Hi all,

During the recent sprint we discussed copy / view semantics in indexing operations. I've tried to summarize *a* proposal that came out of it.

The tldr is

1. Indexing operations always return a view *where possible* (always when indexing columns, if possible when indexing the rows).
2. For consistent behavior, we implement either Copy on Write or Error on Write (explained in the proposal)
3. For readability / understandability, we give users the tools to control in code whether mutations to a "child" dataframe (result of an indexing operation) propagate to the parent, while avoiding unnecessary "defensive" copies.

The full proposal is available at https://docs.google.com/document/d/1csGE4qigPR2vzmU2--jwURn3sK5k5eVewinxd8OUPk0/edit. I'd like to collect comments there and then propose properly by adding it to the roadmap.

cc @pandas-dev/pandas-core

**Comment by jorisvandenbossche** [comment]
cc @pandas-dev/pandas-core

**Comment by jbrockmendel** [comment]
asv from a branch that changes column-based indexing to always return a view:

```
       before           after         ratio
     [42289d0d]       [f99abd57]
     <master>         <myway>   
+       856±100μs       1.28±0.4ms     1.50  index_cached_properties.IndexCache.time_engine('MultiIndex')
-     9.24±0.06ms      8.77±0.02ms     0.95  algorithms.Hashing.time_series_string
-        57.3±2ms       54.3±0.2ms     0.95  join_merge.MergeAsof.time_by_int('backward', 5)
-      76.1±0.3ms       72.0±0.3ms     0.95  frame_methods.Describe.time_dataframe_describe
-        57.3±1ms       54.0±0.2ms     0.94  join_merge.MergeAsof.time_by_int('backward', None)
-     14.0±0.07ms      13.2±0.08ms     0.94  groupby.MultiColumn.time_cython_sum
-        1.29±0ms      1.20±0.01ms     0.94  timeseries.AsOf.time_asof_nan_single('DataFrame')
-     1.50±0.01ms      1.38±0.01ms     0.92  timeseries.AsOf.time_asof_single('DataFrame')
-     3.62±0.01ms      3.31±0.01ms     0.91  groupby.CountMultiDtype.time_multi_count
-      67.0±0.4ms       60.7±0.4ms     0.91  reshape.Crosstab.time_crosstab_normalize_margins
-      14.8±0.4ms      13.2±0.08ms     0.89  join_merge.MergeAsof.time_on_int32('nearest', 5)
-     14.4±0.09ms      12.8±0.04ms     0.89  join_merge.MergeAsof.time_on_int32('nearest', None)
-      14.7±0.1ms      13.0±0.06ms     0.89  join_merge.MergeAsof.time_on_uint64('nearest', 5)
-     14.3±0.05ms      12.6±0.03ms     0.88  join_merge.MergeAsof.time_on_uint64('nearest', None)
-      69.8±0.5ms       61.4±0.7ms     0.88  reshape.PivotTable.time_pivot_table_margins
-     14.3±0.06ms       12.6±0.1ms     0.88  join_merge.MergeAsof.time_on_int('nearest', 5)
-     12.7±0.05ms      11.1±0.04ms     0.87  join_merge.MergeAsof.time_on_int32('forward', 5)
-     14.0±0.06ms      12.3±0.05ms     0.87  join_merge.MergeAsof.time_on_int('nearest', None)
-     12.5±0.05ms      10.9±0.05ms     0.87  join_merge.MergeAsof.time_on_int32('forward', None)
-     12.1±0.05ms      10.5±0.02ms     0.87  join_merge.MergeAsof.time_on_int32('backward', 5)
-     12.0±0.05ms      10.3±0.04ms     0.86  join_merge.MergeAsof.time_on_int32('backward', None)
-     12.7±0.09ms      10.9±0.04ms     0.86  join_merge.MergeAsof.time_on_uint64('forward', 5)
-     12.2±0.06ms      10.5±0.07ms     0.85  join_merge.MergeAsof.time_on_int('forward', 5)
-     12.4±0.03ms      10.6±0.05ms     0.85  join_merge.MergeAsof.time_on_uint64('forward', None)
-     12.0±0.06ms      10.2±0.04ms     0.85  join_merge.MergeAsof.time_on_int('forward', None)
-     11.9±0.05ms      10.1±0.05ms     0.85  join_merge.MergeAsof.time_on_uint64('backward', 5)
-     11.8±0.04ms      9.98±0.05ms     0.85  join_merge.MergeAsof.time_on_uint64('backward', None)
-      11.4±0.4ms      9.61±0.02ms     0.84  join_merge.MergeAsof.time_on_int('backward', None)
-     11.5±0.06ms      9.71±0.04ms     0.84  join_merge.MergeAsof.time_on_int('backward', 5)
-         407±3μs          291±2μs     0.72  reindex.Reindex.time_reindex_columns
-      3.99±0.1ms       2.72±0.2ms     0.68  arithmetic.MixedFrameWithSeriesAxis.time_frame_op_with_series_axis1('ge')
-     3.56±0.02ms      2.36±0.03ms     0.66  arithmetic.MixedFrameWithSeriesAxis.time_frame_op_with_series_axis1('ne')
-      3.60±0.2ms      2.35±0.02ms     0.65  arithmetic.MixedFrameWithSeriesAxis.time_frame_op_with_series_axis1('lt')
-     47.8±0.08ms       28.4±0.1ms     0.59  frame_methods.Dropna.time_dropna_axis_mixed_dtypes('all', 1)
-      4.81±0.9ms      2.81±0.03ms     0.58  arithmetic.MixedFrameWithSeriesAxis.time_frame_op_with_series_axis1('truediv')
-     4.13±0.04ms       2.41±0.2ms     0.58  arithmetic.MixedFrameWithSeriesAxis.time_frame_op_with_series_axis1('eq')
-        11.9±1ms       6.88±0.5ms     0.58  arithmetic.MixedFrameWithSeriesAxis.time_frame_op_with_series_axis1('pow')
-     4.12±0.04ms      2.34±0.05ms     0.57  arithmetic.MixedFrameWithSeriesAxis.time_frame_op_with_series_axis1('le')
-        4.67±1ms       2.55±0.2ms     0.55  arithmetic.MixedFrameWithSeriesAxis.time_frame_op_with_series_axis1('mul')
-        4.90±1ms       2.55±0.2ms     0.52  arithmetic.MixedFrameWithSeriesAxis.time_frame_op_with_series_axis1('sub')
-      40.3±0.2ms       20.5±0.2ms     0.51  frame_methods.Dropna.time_dropna('all', 1)
```

Methodological note: i did a full asv run, then for everything that came back as changed re-ran those and changes that were not reproducible (e.g. tslibs benchmarks which have to be noise)

**Comment by TomAugspurger** [comment]
I'd like to move forward with this, including a version of the proposal on https://pandas.pydata.org/pandas-docs/dev/development/roadmap.html.

There hasn't been much activity on the google doc, https://docs.google.com/document/d/1csGE4qigPR2vzmU2--jwURn3sK5k5eVewinxd8OUPk0/edit. The biggest outstanding item is a decision / discussion of "Copy on Write" vs. "Error on Write". It'd be nice if the proposal was explicit on this point, but I worry we won't be able to decide without actually trying it out.

**Comment by jorisvandenbossche** [comment]
Some assorted thoughts I gathered the last weeks (not necessarily all needing an answer in an initial discussion / proposal, though):

- The copy/view in indexing discussion also impacts methods that rely on indexing internally (eg `drop`, `dropna`). I think that is also shown from the fact that dropna performance improves in the benchmark output above. 
  So the indexing vs methods (what you put in the "extended proposal" section) are certainly somewhat intertwined, although it is of course possible to first only target indexing by ensuring we do additional copies internally in methods like `drop` that expect to return a copy. 

- What to do with duplicate column selection? (`df['a', 'a']]`) Should that also still be a view? (if we would go for views) Mutating one column would then also mutate the other.

- What kind of mutation operations do we consider in this discussion? We clearly talked about "changing values in-place (setitem)", but there are also other kinds of mutation like changing metadata, flags, ..

- What to do with the constructors? (eg `DataFrame(df)`) Also copy-on-write? Or if we keep this as non-copy, what with constructors that have potentially a "reindexing" behaviour? (eg `DataFrame(df, columns=[...])`) 

- What do we do with accessing a single column as a Series? Do we keep this as a pure view, or do we also want to (eventually) introduce copy-on-write/error for this? (after deprecation cycle)

- I think the crucial question that deserves more discussion is: are we fine with ditching the SettingWithCopyWarning for those cases it was introduced for? 
  Because there is a reason that it was introduced: to warn uses that were doing a setitem operation that might have no effect silently, and thus be confusing / easy to miss for users.  

  I think the classical example is something like `df[df['B'] > 4]['B'] = 10` - a chained indexing assignment that doesn't work. The warning was introduced to prevent people doing that and silently having no effect. 
  Are we fine with it that this will indeed have silently no effect? (which would be the consequence of copy-on-write)

**Comment by TomAugspurger** [comment]
Agreed that the classic chained indexing example is a compelling argument for Error on Write rather than Copy on Write.

>  What to do with duplicate column selection? (df[['a', 'a']]) Should that also still be a view? (if we would go for views) Mutating one column would then also mutate the other.

IMO, yes that should create two views to the same data.

> What kind of mutation operations do we consider in this discussion? We clearly talked about "changing values in-place (setitem)", but there are also other kinds of mutation like changing metadata, flags, ..

I'd like to keep this focused just on setting data / values. I'm hopeful that the metadata / flags issue can be solved by doing shallow copies by default and providing an option to copy the data if desired, like we did for `set_flags`.

> What do we do with accessing a single column as a Series?

If possible, treating it the same as a DataFrame seems best to me.

**Comment by jreback** [comment]
is there any system that does error-on-write? I do prefer this, but am a little concerned that we are treading new group vs. copy-on-write (which I *think* has a large body of systems / literature).

**Comment by jorisvandenbossche** [comment]
Putting it here because most of the discussion on this topic is still here (but can also move to https://github.com/pandas-dev/pandas/pull/36988).

While the classic chained indexing example is indeed an argument for Error-on-Write (for this case, it's basically what we have now but with an error instead of warning), I think there are also arguments/examples to give in favor of Copy-on-Write:

* I would personally love to get the (IMO more common case) of `subset = df[mask]` or `subset[list_of_columns]` and then continuing with `subset` without worrying about the original `df` working out of the box, without the user needing to do `.copy()`/`.copy_if_needed()` (as is the case now if you don't want to risk getting the warning, even though those are always already copies)

* There are some cases where we currently don't raise a warning (eg adding a new column to a `subset` created like above), that woud start raising an error.

* We basically need Copy-on-Write anyway, because I think this still is the behaviour we want for *methods*, even if we do Error-on-Write for indexing? (`drop`, `reindex`, ...)

> Also, I've seen the following behavior with people I've
worked with: ...

In this specific case, having consistent Copy-on-Write *could* also "solve" it? If is provides consistent behaviour (always copy, never mutate the passed data), not depending on the actual data passed to the function. 
(of course, it depends on a lot on the actual case, and you were only giving a generic flow. But making things consistent in itself (*always* copy-on-write or *always* error-on-write) could also help avoiding such issues, I think?

**Comment by jorisvandenbossche** [comment]
*Assume* that we could actually detect chained indexing? In this hypothetical situation, would people then be more in favor of Copy-on-Write over Error-on-Write? 

The assumption would be that we can detect the difference in `__setitem__` between

```
df[df['B'] > 4]['B'] = 10
```

vs

```
temp = df[df['B'] > 4]
temp['B'] = 10
```

I am wondering if we shouldn't be able to actually detect this based on reference counting. In the first case, there is no additional reference to `df[df['B'] > 4]` than the calling dataframe (`self`) of the second `__setitem__` operation. 
(I am not sure how robust such a detection would be though)

**Comment by Dr-Irv** [comment]
> In this specific case, having consistent Copy-on-Write _could_ also "solve" it? If is provides consistent behaviour (always copy, never mutate the passed data), not depending on the actual data passed to the function.
> (of course, it depends on a lot on the actual case, and you were only giving a generic flow. But making things consistent in itself (_always_ copy-on-write or _always_ error-on-write) could also help avoiding such issues, I think? (while now you can actually mutate or not, when you get (and ignore) warning)

It's possible a consistent copy-on-write would solve this issue, but we have to make sure we get *all* of the cases right, and I'm just not convinced we'll be able to think all of those cases through.  The other issue for me is that copy-on-write silently does something, which might have side effects that a user would like to be made aware of, whereas error-on-write makes the user very aware of what he is doing.

**Comment by jorisvandenbossche** [comment]
> It's possible a consistent copy-on-write would solve this issue, but we have to make sure we get all of the cases right, and I'm just not convinced we'll be able to think all of those cases through.

Isn't the same true for Error-on-Write? We also need to detect all cases where we need to error instead of allow the mutation.

**Comment by jreback** [comment]
@jorisvandenbossche you have just described the current implementation of SettingCopyWarning and it's entire purpose

it is tracking refs

but i don't actually think this is that robust - and thats the same reason i am not enthusiastic about copy on write

yes if we can guarantee that we can always do it great

but that's the problem i am not sure we really can 

though maybe now that we have full control over PandasArrays it might be fully possible (way back when when i did SettingWithCopy) some ops just indexes directly to the numpy arrays and we had to be very defensive about this

today we have much more robust handling

**Comment by Dr-Irv** [comment]
> > It's possible a consistent copy-on-write would solve this issue, but we have to make sure we get all of the cases right, and I'm just not convinced we'll be able to think all of those cases through.
> 
> Isn't the same true for Error-on-Write? We also need to detect all cases where we need to error instead of allow the mutation.

Yes, I agree.  But if we miss a case, we fix it by raising an exception.  With copy-on-write, we fix it by doing the copy, but I'm guessing that there might be cases where we say "Given this case, we should have chosen error-on-write".

My sense is that it is easier to go from the current state of `SettingWithCopyWarning` to error-on-write and *then* to copy-on-write, than to go from `SettingWithCopyWarning`  to copy-on-write and then discover we have to do  error-on-write  because we simply missed something.  At least as a first step, we can change all the `SettingWithCopyWarning` stuff to just raise an exception and we have a first implementation.  (That's my sense - I could be totally wrong)

**Comment by TomAugspurger** [comment]
Just to clarify the choice, is it fair to say that the only real choice is between

1. Error on Write
2. Copy on Write *with `SettingWithCopyWarning` for chained indexing*?

The initial proposal (and my PR) left out the warning specifically for chained indexing part. *If* we can reliably detect chained indexing (a big if?) then do people prefer option 2?

**Comment by jorisvandenbossche** [comment]
> @jorisvandenbossche you have just described the current implementation of SettingCopyWarning and it's entire purpose

@jreback It might be achieved using the same *mechanism* (I would need to look more into the code of this, it's not a part I am very familiar with), but at least the *goal* I described is different: the current SettingWithCopyWarning warns in both example cases, while I want only the first (actual chained indexing) to raise, and not the second (assigning the subset to a variable).

> Yes, I agree. But if we miss a case, we fix it by raising an exception. With copy-on-write, we fix it by doing the copy, but I'm guessing that there might be cases where we say "Given this case, we should have chosen error-on-write".

@Dr-Irv I don't fully understand your last sentence.  
It's true that, in case of a bug in the detection code when to copy/error, the fix is more explicit with error-on-write ("loud", since people will start seeing an error if they relied on wrong behaviour, instead of simply no longer getting the wrong behaviour). 
So do you mean that even if we choose copy-on-write, we need to fix cases by raising an error instead of copying because it would otherwise be a "silent" fix?

**Comment by jorisvandenbossche** [comment]
> Just to clarify the choice, is it fair to say that the only real choice is between

Not sure if there is anybody in favor of it, but the original write-up also still has the option of "clear rules" (eg indexing columns could *always* be *actual* views)

**Comment by jbrockmendel** [comment]
> Not sure if there is anybody in favor of it, but the original write-up also still has the option of "clear rules" (eg indexing columns could always be actual views)

I prefer this over the status quo.

**Comment by Dr-Irv** [comment]
> @Dr-Irv I don't fully understand your last sentence.
> It's true that, in case of a bug in the detection code when to copy/error, the fix is more explicit with error-on-write ("loud", since people will start seeing an error if they relied on wrong behaviour, instead of simply no longer getting the wrong behaviour).
> So do you mean that even if we choose copy-on-write, we need to fix cases by raising an error instead of copying because it would otherwise be a "silent" fix?

No, I mean let's say that we choose copy-on-write.  We think we get it right.   We release the code.  Someone discovers a place where we didn't get it right, and then we discuss it, and realize we should have chosen "error-on-write" in the first place.  I'm just erring on the side of caution by preferring error-on-write.

**Comment by jorisvandenbossche** [comment]
>  let's say that we choose copy-on-write. We think we get it right. We release the code. Someone discovers a place where we didn't get it right, and then we discuss it, and realize we should have chosen "error-on-write" in the first place. I'm just erring on the side of caution by preferring error-on-write.

OK, understand it now ;) 
But, I personally think we should discuss / choose what we *actually* prefer. Let's assume that after more discussions we think we prefer copy-on-write over error-on-write, then I would say we should go with that, even if that makes it more difficult to still change mind afterwards. 
This also touches upon implementation / adoption questions which we didn't really discuss yet. I think we will need *some* way to opt-in to this new behaviour initially / have a flag to enable it / .. as an experimental stage in which we can still make breaking changes (like going from copy-on-write to error-on-write, which in principle could also still be done with deprecations first). This might be quite complex to first have both old and new behaviour side by side, but not sure there is a good alternative ..

(and to be clear: while I am mainly arguing here for Copy-on-Write to give some counter-arguments, I am not yet sure myself which of the two is "best". I mainly think it's something to futher explore/experiment/discuss)

> > Not sure if there is anybody in favor of it, but the original write-up also still has the option of "clear rules" (eg indexing columns could always be actual views)
>
> I prefer this over the status quo.

I also prefer that over status-quo. But, @jbrockmendel, do you also prefer that over the copy-on-write/error-on-write alternatives discussed? (personally, I *think* that I am currently in favor of one of those two other options)  
Note that also in this case of "clear rules", the discussion around what to do with the "typical chained indexing example" is still needed (eg if we have the rule that subsetting rows is always a copy, do we stop raising SettingWithCopyWarning for cases like `subset = df[mask]` with subsequent changes ?)

**Comment by Dr-Irv** [comment]
Here's something that I brought up in the October, 2020, dev meeting.  Consider:
```python
>>> df = pd.DataFrame({"x":[1,2,3], "y":[4,5,6]})
>>> df
   x  y
0  1  4
1  2  5
2  3  6
>>> sx=df['x']
>>> sx.iloc[1] = 10
>>> sx
0     1
1    10
2     3
Name: x, dtype: int64
>>> df
    x  y
0   1  4
1  10  5
2   3  6
>>> sdf = df[['x']]
>>> type(sdf)
<class 'pandas.core.frame.DataFrame'>
>>> sdf
    x
0   1
1  10
2   3
>>> sdf.loc[1,'x'] = 20
C:\Anaconda3\lib\site-packages\pandas\core\indexing.py:670: SettingWithCopyWarning:
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)
__main__:1: SettingWithCopyWarning:
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
>>> sdf
    x
0   1
1  20
2   3
>>> df
    x  y
0   1  4
1  10  5
2   3  6
```
Right now, the behavior of picking a single column using the syntax `df['x']`, which creates a `Series` versus `df[['x']]` which creates a `DataFrame` is different, in that when you get a `Series`, modification of that series silently modifies the underlying `DataFrame` while when you get a `DataFrame`, you get the `SettingWithCopyWarning`.  If we choose copy-on-write, then someone who used the `df['x']` syntax with the dependence that the underlying `DataFrame` was modified will no longer have working code, and will have a hard time figuring out that a new version of pandas is what caused the issue.

On the other hand, if we use error-on-write, then in BOTH cases, we raise an error, which makes it clear to the user that they need to make their intent clear.

**Comment by jbrockmendel** [comment]
Do either the CoW or EoW allow for the possibility that when i edit a view I am doing so intentionally?

**Comment by Dr-Irv** [comment]
> Do either the CoW or EoW allow for the possibility that when i edit a view I am doing so intentionally?

By "doing so intentionally", you mean modifying the underlying `DataFrame` as a result of modifying the view?

**Comment by jbrockmendel** [comment]
> By "doing so intentionally", you mean modifying the underlying DataFrame as a result of modifying the view?

Yes.  A case where I want to modify both.

**Comment by Dr-Irv** [comment]
> > By "doing so intentionally", you mean modifying the underlying DataFrame as a result of modifying the view?
> 
> Yes. A case where I want to modify both.

Yes, the proposal has a `as_mutable_view()` method to allow the user to indicate they want to allow mutation.  Works whether we do COW or EOW.

**Comment by toobaz** [comment]
> Not sure if there is anybody in favor of it, but the original write-up also still has the option of "clear rules" (eg indexing columns could _always_ be _actual_ views)

Sorry, I'm missing the difference between these "clear rules" and the "whenever possible", could you clarify?

I'm also missing one thing in the proposal: how can something like ``df[["A"]].as_mutable_view()`` be implemented? Either ``df[["A"]]`` already provides a mutable view - hence the method call is noop - or it provides a copy of data - and it will be painfully hard to keep track and edit the original data (right?).

A syntax I would understand would be something like ``df.v[["A"]]`` - where ``.v`` is lazy and only operates when indexed. But even leaving aside the syntax, are we thinking this should work for _all_ indexing operations?! For instance ``df.loc[['a', 'b'], slice(1, 10, 4)]``?! Not that it's impossible, but it does looks difficult to me.

**Comment by TomAugspurger** [comment]
> I'm also missing one thing in the proposal: how can something like df[["A"]].as_mutable_view() be implemented?

Indexing columns would always give views. The difference comes when you try to mutate the view

```python
>>> df2 = df1[["A']]  # df2 is a view on df1
>>> df2.iloc[:, :] = 0  # Copies (with Copy on Write) or raises (with Error on Write).
``` 

And for the case of "I want to mutate `df1`, you'd use `df2 = df1[["A"].as_mutable_view()`.

> are we thinking this should work for all indexing operations?

I think this discussion is limited to

- Indexing just the columns
- (maybe) Indexing the rows in a way where NumPy can return a view when slicing a 1D array. So `df.loc[:5, ["A"]]` can be a view. But your example of ``df.loc[['a', 'b'], ["A"]]` cannot be a view, and would copy immediately.

**Comment by toobaz** [comment]
> Indexing columns would always give views. The difference comes when you try to mutate the view

Yep, I had completely misunderstood, now I got it, as well as the difference between the "clear rules" and the "whenever possible": that the former would stay more or less with the status quo of creating mutable views.

If I now got it right, I still have one question: what makes us different from numpy, where the status quo is sensible and does not seem to be under discussion?

Is it because our users have different needs, or are less smart? I don't think so.

Is it because our implementation is different? The only relevant difference I see is that any indexing on columns could return views... but this is still a "clear rule", not particularly difficult to grasp.

So my feeling is that ``SettingWithCopyWarning`` was born only because our rules were not clear, and that this is a problem we can fix, and that the proposal is significantly more complicated for users to understand.

In the long run, we could remove the warning - or make it optional - _and_ provide a way to set the CoW and/or EoW behavior. All of this should not require more code than what is being discussed here, but it would not change the default behavior. And by the way EoW would be annoying in many cases (including the internals) as a default, but it would be a great tool as an optional.

This said, I do acknowledge that this was discussed live, so I hope I'm not saying something completely obvious - and if I am, please feel free to tell me.

Coming back to the syntax, however,

> And for the case of "I want to mutate `df1`, you'd use `df2 = df1[["A"].as_mutable_view()`.

I think this is dangerous. ``as_mutable_view()`` is something that basically does not affect ``df2`` (with CoW), or affects it only in the sense of not raising errors (with EoW), while it does affect ``df1`` (because it allows its content to change). This is not intuitive.

If I currently write ``user_provided_method(df1[["A"]])`` I know (because it is the default) that ``user_provided_method`` might change ``df1``. But if the default becomes CoW or EoW, I might expect ``df1`` to be safe. Instead the user can still call ``as_mutable_view()``. So either I do ``user_provided_method(df1[["A"]].copy(deep=True))`` (possibly wasting a copy), or we establish a hierarchy by which ``copy_if_needed()`` wins over a later ``as_mutable_view()`` (and assume that users know both, and their hierarchy). Isn't it much cleaner to have ``df2 = df1.mutable[["A"]]`` (lazy), where mutability cannot be applied retroactively (at least through the API)?

**Comment by jbrockmendel** [comment]
> what makes us different from numpy, where the status quo is sensible and does not seem to be under discussion?

The status quo is not sensible.  e.g.

```
arr = pd.array([1, 2, 3])   # <- IntegerArray
df = pd.DataFrame({"A": arr})
df2 = df[["A"]]

df3 = df.astype("i8")
df4 = df3[["A"]]
```

`df2` is a view on `df`, while `df4` is a copy of `df3`.

Sensible behavior would be either always-copy or always-view.  I maintain

a) there isn't a compelling reason to treat `df[["A"]]` differently from `df["A"]`
b) "indexing on columns always gives views" is easier to document/explain/grok than any mixed alternative
c) views are better for perf.

**Comment by toobaz** [comment]
> The status quo is not sensible. e.g.

Sorry, I was referring to the numpy status quo, presumably amended with "any indexing on columns returns views". So it would be "a (mutable) view is always returned except when rows are indexed in such a way that would return a copy in numpy [with lists/slices of labels being assimilated to lists/slices of positions]". I'm all for stating and enforcing clear rules with selected API changes, while I'm afraid that the more drastic changes to default behavior (CoW or EoW) are not what users expect.

**Comment by TomAugspurger** [comment]
> If I now got it right, I still have one question: what makes us different from numpy, where the status quo is sensible and does not seem to be under discussion?

I think the primary difference is that DataFrames are "more columnar" than ndarrays. I would claim that

1. Slicing a subset of the columns `df[['A', 'B']]` is more common than slicing a subset of the second axis of an ndarray ``ndarray[[:, [0, 1]]`
2. Our data structure allows for more things to be views (at the "cost" of less consolidation). In NumPy any kind of fancy indexing / boolean masking results in a copy. We can support fancy indexing (integer or label-based) or boolean masks on the columns returning views.

**Comment by toobaz** [comment]
> I think the primary difference is that DataFrames are "more columnar" than ndarrays.

I totally agree. To my eyes, your two points are precisely what justifies the "any indexing on columns returns views" variation - not the CoW or EoW as default.

**Comment by jorisvandenbossche** [comment]
Regarding the example of @Dr-Irv at https://github.com/pandas-dev/pandas/issues/36195#issuecomment-708605199:

> If we choose copy-on-write, then someone who used the `df['x']` syntax with the dependence that the underlying DataFrame was modified will no longer have working code, and will have a hard time figuring out that a new version of pandas is what caused the issue.

If we choose copy-on-write, and also use this for Series access, we for sure need to raise warnings about that this will no longer happen in the future / is no longer happening for a long time, I think.  Because it's indeed true that this would be a tricky case to upgrade for (but with good warnings, still doable IMO). 
But, whether`df['x']` will stop being a view is still something we need to discuss / decide. It's not explicitly mentioned in the proposal, and also the reason I raised this question above https://github.com/pandas-dev/pandas/issues/36195#issuecomment-702195996. 

>> [@toobaz ] I'm also missing one thing in the proposal: how can something like df[["A"]].as_mutable_view() be implemented? 
>>  are we thinking this should work for all indexing operations?
>
> [@TomAugspurger] I think this discussion is limited to
>
> * Indexing just the columns
> * (maybe) Indexing the rows in a way where NumPy can return a view when slicing

To add to this: `as_mutable_view()` will "work" (in the sense of being *available*) for all indexing operations. But how it "works" is that when the subset is actually a copy already (so mutating the parent is not possible), the method will raise an error saying that it cannot give you a mutable view.  
In which cases it won't raise an error, that's indeed the list that Tom provided.

> [@toobaz, about `df2 = df1[["A"].as_mutable_view()`] I think this is dangerous. `as_mutable_view()` is something that basically does not affect `df2` (with CoW), or affects it only in the sense of not raising errors (with EoW), while it does affect `df1` (because it allows its content to change). This is not intuitive.  
>
> If I currently write `user_provided_method(df1[["A"]])` I know (because it is the default) that user_provided_method might change df1. But if the default becomes CoW or EoW, I might expect `df1` to be safe. Instead the user can still call `as_mutable_view()`. So either I do `user_provided_method(df1[["A"]].copy(deep=True))` (possibly wasting a copy), or we establish a hierarchy by which `copy_if_needed()` wins over a later `as_mutable_view()` (and assume that users know both, and their hierarchy). Isn't it much cleaner to have df2 = df1.mutable[["A"]] (lazy), where mutability cannot be applied retroactively (at least through the API)?

I would say that it is the responsibility of the function to document whether it will mutate input or not, but indeed, you can of course not be sure about this (like it is the case now as well). 
Note that *if* we do CoW or EoW, we would have a copy method that only "copies if needed", to avoid the "possibly wasting a copy". This is actually also one of the main drivers (at least for me) to get away from "status quo". As currently you need to waste a copy when eg doing boolean indexing, where you know you already have a copy of the original dataframe, but need to do another copy to avoid getting warnings (but to be clear: the "clear rules" option without the current warnings would also solve that).

Having a way to get this mutable view that doesn't do this retroactively (like your `df2 = df1.mutable[["A"]]`) is an interesting idea (if we would opt for EoW/CoW). That might indeed give more explicit control over mutability to the "owner" of the dataframe (so, if you as a user didn't make the subset mutable, a function receiving your dataframe also won't be able to do it).

**Comment by toobaz** [comment]
> Note that _if_ we do CoW or EoW, we would have a copy method that only "copies if needed", to avoid the "possibly wasting a copy".

Yes, we just need to make sure/clear this cannot be overridden by a later ``.as_mutable_view()``.

**Comment by jorisvandenbossche** [comment]
I think it would be useful to work out a bit more what the option of **"clear rules"** could look like, to be able to better compare it to CoW or EoW. 
(because also here there are still different options, and open questions about what to do with warnings ..)

---

One possible set of well-defined copy/view rules for indexing:

* Subsetting (one or more) columns is always a view
* Subsetting rows is always a copy (so subsetting both rows and columns is also a copy)

One potential addition to the above two rules could be:

* Subsetting rows with a *slice* is always a view (so the combo of slicing rows and subsetting columns is also a view)

With this addition, we would basically follow numpy rules for row selection, but have our own rule (always a view) for columns, since we can do this for any type of column selection given the columnar nature of a DataFrame. 
But the additional also complicates the rules a bit more (at least for people not familiar with numpy's rules, but I think this is a large part of our user base).

Also in this case, the question comes up of what to do with the "classical chained indexing" example:

```python
# raise a warning or not?
# According to our "clear rules" this will not update `df`, but it still remains a
# gotcha like it always has been (the reason we actually have this warning)
df[df['B'] > 4]['B'] = 10

# raise a warning or not? 
# For me one of the reasons for this whole discussion is that the below can be done (in some way) without warning
subset = df[df['B'] > 4]
subset['B'] = 10
```

--- 

A different possible set of well-defined copy/view rules for indexing:

* Getting a single column as a Series is always a view
* Anything else is always a copy (so any indexing operation that results in a DataFrame and/or subsets rows)

And again there could be an additional rule for slicing rows (although here that would only apply for slicing rows while selecting a single column as a Series, then)

This is another set of "clear rules" that I think is closer to current behaviour (so you could see it as a "rationalization" of the current behaviour). 
BTW, returning a "copy" for selecting multiple columns doesn't necessarily mean we need to return a physical copy, as long as it appears a copy to the user, to follow those "clear rules". Although with this note, this is actually a variant of CoW ..

Also here, the same question about warning or not for `df[df['B'] > 4]['B'] = 10` applies.

**Comment by toobaz** [comment]
> With this addition, we would basically follow numpy rules for row selection

Indeed this is what I had this in mind, but I had overlooked that in numpy selecting a single row results in a view, which is not something we want, as long as in the long term we don't expect pandas to support 2D data blocks (a ``Series`` should really not have to look for its data scattered in N different 1D blocks).

So if this long term assumption holds, I would be uncertain between indexing rows
- to always return a copy, or
- to return a copy except with ``slice``s (but not with single labels/positions)

... and in the end I would have a slight preference for the former as long as the latter is very easy to do (e.g. ``as_mutable_view``, whatever the syntax).

> Also here, the same question about warning or not for `df[df['B'] > 4]['B'] = 10` applies.

To my eyes, the entire discussion on chained indexing is the consequence of not having clear rules. If we do have clear rules _and_ a handy way to choose between copies and views, we do not need to warn users.

My preferred solution is:
- ``df.loc``: "clear rules"
- ``df.v.loc``: a view whenever possible, raise otherwise
- ``df.c.loc``: CoW
Users who want to play it safe use ``df.c.loc``. Users who look at performance can use ``df.v.loc`` to avoid undetected copies. Users who don't have special needs (who I think are the majority) will go on as always.

In the docs, we'll just warn users that chained indexing like `df[df['B'] > 4]['B'] = 10` is not guaranteed to work, but if they want to make sure it works (where possible), they can do `df.v[df['B'] > 4].v['B'] = 10` (unless we make ``.c`` and ``.v`` inheritable - in the spirit of the proposal - so that `df.v[df['B'] > 4]['B'] = 10` is enough, but we need a third accessor to go back to "normal mode").

**Comment by Dr-Irv** [comment]
> My preferred solution is:
> 
> * `df.loc`: "clear rules"
> * `df.v.loc`: a view whenever possible, raise otherwise
> * `df.c.loc`: CoW
>   Users who want to play it safe use `df.c.loc`. Users who look at performance can use `df.v.loc` to avoid undetected copies. Users who don't have special needs (who I think are the majority) will go on as always.

I like the concept of having explicit methods for getting views.  (e.g. `df.v.loc`).  

The issue is what to do with `__getitem__()`.  @toobaz what rules would you have if `.loc` was not being used, e.g. `df['A']`, `df[['A']]`, `df[['A','B']]`, `df[df['A']>3]['B']`, `df[df['A']>3][['B']]` ?

So I'll throw out another idea, which is that we leave existing behavior as it is today, but we implement the code from @jbrockmendel that returns views of a `DataFrame` as a new method on `DataFrame`.  So if you want performance, you do `df.view[['A','B']]` and that gives you a view, but `df[['A','B']]` gives a copy as it does today.   Then the following sequence would give the `SettingWithCopyWarning`:
```
dfv = df.view[['A','B']]
dfv.iloc[0,'A'] = 10  # Yields warning - it's a view
```

**Comment by jorisvandenbossche** [comment]
> Then the following sequence would give the SettingWithCopyWarning:

That doesn't enable the usecase of "I want a view so I can mutate inplace", though. If you explicitly ask for a view, I would expect to *not* get a warning? (the user knows they have a view, so should also know that mutating the subset mutates the parent?)

> [@toobaz] we'll just warn users that chained indexing like `df[df['B'] > 4]['B'] = 10` is not guaranteed to work, but if they want to make sure it works (where possible), they can do `df.v[df['B'] > 4].v['B'] = 10`

That specific example cannot actually work. A boolean mask always gives a copy, so you're mutating a column of a copy. 
(unless we would start with "lazy filtering", keeping track of the selection filter, without doing the actual selection until some next operation. A very interesting topic, but we probably already have enough on out plate in this discussion .. ;))

**Comment by toobaz** [comment]
> what rules would you have if `.loc` was not being used,

... depends on what rules we settle on, but..

> e.g. `df['A']`

same as `df.loc[:, 'A']`

> `df[['A']]`

same as `df.loc[:, ['A']]`

> `df[['A','B']]`

same as `df.loc[:, ['A', 'B']]`

>`df[df['A']>3]['B']`

same as `df.loc[df['A']>3].loc[:, 'B']`

> `df[df['A']>3][['B']]`

same as `df.loc[df['A']>3].loc[:, ['B']]`

I *think* these identities should be valid already and should stay valid whatever decisions we take. Also because the rules shouldn't make any reference to `loc` vs. `_getitem_` vs. `iloc`... only to type of indexer and/or dimensionality of the result.

**Comment by toobaz** [comment]
> That specific example cannot actually work

Correct! I took a bad example

**Comment by Dr-Irv** [comment]
> That doesn't enable the usecase of "I want a view so I can mutate inplace", though. If you explicitly ask for a view, I would expect to _not_ get a warning? (the user knows they have a view, so should also know that mutating the subset mutates the parent?)
> 
> >

That's new functionality, but today, you can mutate in place if you grab a single `Series`.  So we would still allow that kind of mutate-in-place.  We just wouldn't allow the *new* functionality of getting a view that returns a `DataFrame` and mutating that view, but that doesn't work today anyway.

What I am suggesting here (just create a new method to return a view of a `DataFrame` when asking for multiple columns) would then allow us to preserve all the existing behavior, get the benefits of performant views of a `DataFrame` when you want it, and where you don't need to modify it, and might be an incremental step towards the eventual behavior where we have "rules" that determine whether a view or copy is returned using the existing operators.

**Comment by jorisvandenbossche** [comment]
I would like to revive this discussion.

From thinking a bit more about this the last months (and re-reading the discussion above), I think I am currently most convinced of the following concrete proposal:

* Any indexing operation (subsetting a DataFrame or Series in any way) always returns a copy in terms of user API.
* Under the hood we actually use views as much as possible and use Copy-on-Write to ensure the user API of always copying.

In addition, we can still have some way to get an actual view (whether it is the `as_mutable_view()` from the original proposal, or another indexer like `df.view[..]` or `df.v.loc[..]`, that's still to be seen). 

Another addition: I think we can also extend this to *creation* methods (creating a DataFrame/Series from existing data, eg `DataFrame({'col': arr_or_series})`, or adding columns like `df["new_col"] = arr_or_series`), where the resulting DataFrame/Series would by default also never modify the original using C-o-W (and for constructors this could be overriden with an explicit `copy=False`).

Motivations:

* If we also want to improve the copy/view efficiency of *methods* (eg `reset_index`, `drop` to drop columns, `rename`, etc) we will need to implement a form of C-o-W anyway, I think (Error-on-Write would not be appropriate for those cases IMO). Personally, I find this an important improvement that I would like to see happen. 

* This proposal can be seen as a "clear rule", but using C-o-W under the hood as an implementation detail to delay the actual copy until it is needed. 
  And when it comes to clear rules, I think the rule of "always copy" is the simplest "clear rule" we can get. 
  Other "clear rule" ideas above would always include some special cases (and deviations from the numpy rules). And even with clear rules a user still needs to know the details of those rules to understand that `df['a'][df['b'] < 0] = 0` or `df[df['b'] <0]['a'] = 0` (switched order of column/row indexing) does something differently. 
  

* Assignment with chained indexing will basically never work in this case, which makes that we can always warn, or actually even error (to be very explicit) for it. This would improve the chained indexing pitfall situation a lot, I think.
  (to be clear, this argument applies to E-o-W as well, but not necessarily on other "clear rules" ideas).

* One of my main motivations is that filtering a dataframe like `subset = df[df["col"] == val]` just works out of the box without the user needing to take additional steps to avoid copies / warnings (so eg not needing to do `subset = subset.copy()` to avoid warnings (current situation) or errors (with Error-on-Write)).

* My general feeling is that Copy-on-Write tailors to the majority of the use cases where you don't actually want to (or care about) modify the "parent" dataframe, while Error-on-Write is a bit safer (more explicit) for the use cases where the user actually wants to modify the parent (but which might silently not happen anymore with Copy-on-Write), but at the cost of making the majority of use cases less convenient. 
 Some specific arguments to choose C-o-W in my opinion: 
  1) I think the first set of use cases (don't modify parent dataframe) is currently by far the most frequent, and I would like to make this use case as convenient as possible (so I would prefer that it is not needed for the user to "make their intent clear" by eg `copy_if_needed()`) 
  2) To smooth a transition for such a breaking change, I think that we can have a version of pandas that warns / errors in all cases that would now modify the parent but would no longer do that in the future (and eg we could first have a pandas 2.0 with all those warnings/errors, and then a pandas 3.0 that does the actual switch). 
  3) The cases where you now actually *can* modify the parent dataframe is quite limited, and I see two main cases: 1) slicing the rows of a DataFrame and 2) selecting a single column. Personally, I have never done something like `subset = df[lab1:lab2]` with the intent of wanting to modify the parent `df` through `subset`, so IMO we shouldn't base our decision on such a case. The second case of selecting a single column is, I suppose, much more frequent. We could think about making this safer by eg warning/erroring for a longer time for this specific case?

**Comment by toobaz** [comment]
> I think I am currently most convinced of the following concrete proposal:

I like it. Just one objection, which however does not undermine my support for your proposal:

>     * Assignment with chained indexing will basically never work in this case, which makes that we can always warn, or actually even error (to be very explicit) for it

My understanding is that warnings/errors would still be based on some sort of counting references, the implementation of which would not be significantly simplified. But I don't see this as a big stopper - after all, in Python, ``[1, 2, 3, 4][2:4].append(3)`` doesn't emit any error or warning.

**Comment by jbrockmendel** [comment]
> we will need to implement a form of C-o-W anyway

Assuming away complexity/perf concerns, something like this does sound like our best option.  Do we have anything like a POC, or anything on the horizon?

If we aren't likely to implement CoW in the near-future, we shouldn't let the perfect be the enemy of the good.

**Comment by jorisvandenbossche** [comment]
>  Do we have anything like a POC, or anything on the horizon? If we aren't likely to implement CoW in the near-future, we shouldn't let the perfect be the enemy of the good.

I am doing some small experiments at the moment, and if there is some level of agreement on the way to go, I am planning to work on it in the coming month.

**Comment by jorisvandenbossche** [comment]
> > * Assignment with chained indexing will basically never work in this case, which makes that we can always warn, or actually even error (to be very explicit) for it
> 
> [@toobaz] My understanding is that warnings/errors would still be based on some sort of counting references, the implementation of which would not be significantly simplified. But I don't see this as a big stopper - after all, in Python, `[1, 2, 3, 4][2:4].append(3)` doesn't emit any error or warning.

The implementation will indeed involve some sort of reference counting, but might not necessarily be the same as we have now. I explicitly want to distinguish between:

```
subset = df[df['b'] <0]
subset['a'] = 0
```
and

```
df[df['b'] <0]['a'] = 0
```

where only the second would raise a warning/error (while currently both warn). Based on a quick experiment some time ago, I think it is doable to detect specifically the second case (which is a case where we know that 1) the user *intents* to modify `df` while 2) `df` will actually never be modified). 

It's true that modifying objects with chained indexing eg with Python lists also doesn't work. But since the above actually worked for a long time in pandas (if the row / column indexer is reversed), I think it would be good to explicitly warn/error for this case.

**Comment by jorisvandenbossche** [comment]
Gentle ping for this discussion. Are there more thoughts on my proposal for "always copy through copy-on-write" at https://github.com/pandas-dev/pandas/issues/36195/#issuecomment-786654449 ?

**Comment by jbrockmendel** [comment]
> I am doing some small experiments at the moment, and if there is some level of agreement on the way to go, I am planning to work on it in the coming month.

Any progress to report?  If not, my thoughts remain roughly https://github.com/pandas-dev/pandas/issues/36195#issuecomment-786740079

**Comment by jorisvandenbossche** [comment]
One aspect I am running into while working on a proof-of-concept implementation is: **can we still have "shallow" copies?** (aka what will `df.copy(deep=False)` do?)

Currently, when making a copy of a DataFrame ([`DataFrame.copy()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.copy.html#pandas-dataframe-copy)), it will make a "deep" copy by default (copy the actual data of the columns), but you have the option of `deep=False` to have a "shallow" copy: a new DataFrame object (with a new BlockManager object) but referencing the same arrays:

> When `deep=False`, a new object will be created without copying the calling object’s data or index (only references to the data and index are copied). Any changes to the data of the original will be reflected in the shallow copy (and vice versa).

This means you can have behaviour like this:

```python
>>> df = pd.DataFrame({'a': [1, 2, 3], 'b': [.1, .2, .3]})
>>> df2 = df.copy(deep=False)

# mutating existing columns of df2 also updates original df
>>> df2.iloc[0, 0] = 100
>>> df
     a    b
0  100  0.1
1    2  0.2
2    3  0.3

# but df2 is a separate BlockManager, so adding new columns/rows doesn't update original df
>>> df2['c'] = 1
>>> df
     a    b
0  100  0.1
1    2  0.2
2    3  0.3
```

However, I am not sure we can reliably preserve such functionality when also having Copy-on-Write. 

Assume the following case:

```python
df = pd.DataFrame(...)
subset = df[1:3]
subset2= subset.copy(deep=False)
```

The `subset` dataframe actually *views* the same data as `df` because it was constructed using a slice (but it keeps track of that reference, so with C-o-W, mutating `subset` or `df` will trigger a copy). But now I created a "shallow" copy of `subset`. According to the current logic, mutating `subset2` should also update `subset` (and vice versa), but should not update `df`. 

Currently, I am simply keeping track of references (with `weakref`), so that we know that we have to copy an array on setitem if there is any other object also referencing that array. However, to support the above logic, we would also need to keep track of a second type of references for cases where it's actually a view for the user (though shallow copy), and thus both objects need to be kept synchronized. 
(concrete example: when mutating `subset` (eg `subset.iloc[0, 0] = 100`), C-o-W will trigger a copy of the array for that column, because it references data of `df`. But if we want to keep the "shallow" relationship between `subset` and `subset2`, this setitem operation on `subset` should also replace that array in `subset2`)

Adding such logic will certainly make it a lot more complicated, but I am also not sure it can even be done reliably (and also adds new dubious corner cases, e.g. what if I added a column to `subset2` in the meantime, are the other columns still views on the columns of `subset`?) 

**Are there actually use cases where `df.copy(deep=False`) is useful?** And then I mean specifically the aspect of it that mutations are propagated, because if you don't need this aspect, C-o-W already gives you the aspect of "new dataframe object without copying the data". 

So maybe we can change the "shallow" copy idea to mean the "C-o-W delayed copy" concept, so that `df2 = df.copy(deep=False)` would be equivalent to `df2 = df[:]` ? (new object referencing same underlying data, but mutations are not propagated)

**Comment by jbrockmendel** [comment]
These are some tough issues, thanks for taking a look at them.

> The subset dataframe actually views the same data as df because it was constructed using a slice (but it keeps track of that reference, so with C-o-W, mutating subset or df will trigger a copy). But now I created a "shallow" copy of subset. According to the current logic, mutating subset2 should also update subset (and vice versa), but should not update df.

The example you gave `subset = df[1:3]` is slicing along rows.  Does anything become easier if you focus on CoW for only indexing along columns?  The existing behavior for slicing on rows is simple+predictable, so if it is appreciably easier to implement CoW only for columns, that would be worth considering.

> Are there actually use cases where df.copy(deep=False) is useful? And then I mean specifically the aspect of it that mutations are propagated

Nothing comes to mind.

**Comment by jorisvandenbossche** [comment]
> Does anything become easier if you focus on CoW for only indexing along columns? The existing behavior for slicing on rows is simple+predictable, so if it is appreciably easier to implement CoW only for columns, that would be worth considering.

Based on my current understanding / implementation, it wouldn't help. Consider the following case:

```python
df = pd.DataFrame(...)
subset = df[1:3]
subset2= subset[['col1', 'col2']]
```

Here, when limiting CoW for only column-indexing, `subset` would be an actual view on `df`, while `subset2` is a CoW-view on `subset`. This means that when modifying `subset2`, it shouldn't modify the parent `subset`, but also the other way around: modifying the parent `subset` should not modify the child `subset2`. To achieve this, CoW would ensure we copy the column  array in `subset` when modifying it. But that would break the view-relationschip between `df` and `subset`. 

Basically, when implementing this in pandas in python (it's not a language-level feature), I don't think we can reliably mix real views and CoW views between pandas objects (DataFrames / Series). So I *think* that for CoW, it would be all or nothing.

**Comment by nickeubank** [comment]
LOVE This. Just to flag I tried a version of this back in the day, in case anything I played with is helpful for implementation. I think constructors is where I really got into messes. https://github.com/pandas-dev/pandas/pull/11500. But this is a huge pain point for students (and instructors! :)) I'd love to see fixed.

**Comment by jorisvandenbossche** [comment]
@nickeubank yes, I went through your PRs and the discussion from back when before starting my proof-of-concept implementation (I am also using weakrefs, although with a slightly different approach: at the array level inside the manager instead of at the DataFrame level). Thanks for your work from back then!

Giving the PR a test drive is certainly welcome.

**Comment by jorisvandenbossche** [comment]
I created a modified version of Tom's google doc with a longer version of the proposal: https://docs.google.com/document/d/1ZCQ9mx3LBMy-nhwRl33_jgcvWo9IWdEfxDNQ2thyTb0/edit?usp=sharing, and also notified the mailing list about this proposal.

**Comment by nickeubank** [comment]
> @nickeubank yes, I went through your PRs and the discussion from back when before starting my proof-of-concept implementation (I am also using weakrefs, although with a slightly different approach: at the array level inside the manager instead of at the DataFrame level). Thanks for your work from back then!

Haha -- glad you found something useful in there! As you can still I got quite stuck, so I'm glad you found a workaround by embedding the functionality at the array manager level.

**Comment by jorisvandenbossche** [comment]
One aspect that has come up on the mailing list, is the risk that users can incorrectly mutate views in places where we expose internal data. 

Since this is Python, a user will always find some (private) way to incorrectly mutate data without triggering the copy-on-write paths, but we should indeed look into ways to protect against such improper mutation of views.

Listing the possible ways to get the internal "array" data from DataFrame/Series objects:

* `Series.values` / `Series.array` -> returning a numpy array or pandas ExtensionArray, which currently return the stored data as mutable arrays as is (or as views). Mutating such an array wouldn't trigger Copy-on-Write which is managed on the DataFrame/Series level. To prevent users from doing this, we could return those arrays as "read-only"? (to avoid always doing a defensive copy here)
  * (numpy arrays already have a concept of "read-only", for EAs we would need to add this)

* `Series.to_numpy()` -> returning a numpy array. This method has a `copy` keyword with currently a default of False. We could either make this copy=True by default, or similarly to the above make it read-only by default, leaving the copy=True/False options to choose from explicitly.

* `DataFrame.to_numpy()` / `DataFrame.values` -> returning a 2D numpy array, which is by definition always a copy (by concatting multiple 1D arrays). Except for the 1-column case, this could still be a view. For simplicity, I would make this case return a copy as well (if you want a view the user can get the Series). Or alternatively this case could follow the logic of Series.to_numpy above.

**Comment by nickeubank** [comment]
As memory serves, numpy arrays moving in and out of pandas was definitely where I got stuck in my own implementation attempt. I think I struggled in particular with people passing numpy arrays into `pd.DataFrame()` without a defensive copy, but I think the situation here is analogous.

Do we think it'd be better for `Series.values` to return a read-only copy, or just return a copy? The former is faster, but the latter seems far more user-friendly. Personally, given anyone who really needs performance would be able to get a view using `Series.to_numpy()`, and I think for most users that performance hit isn't that big of a deal, my inclination would be just to do a defensive copy for `Series.values`. 

Moreover, that would prevent a breaking change, since I imagine many people have `Series.values` objects that they manipulate in their current code, and that would break if we moved to returning "read-only" arrays.

**Comment by jreback** [comment]
I would like to see what happens in our test suite if we change *both* `.to_numpy()` and `.values` to return read-only view as per https://github.com/pandas-dev/pandas/issues/36195#issuecomment-881405656 . I view this as a weird contract that we have with the outside world, e.g. a private api essentially. Ideally would like to remove this access point. Yes we *could* copy as per @nickeubank but ok with doing the api change i think. Of course this is basically impossible to deprecate so would have to make a decision on this and release in a major update.

**Comment by jbrockmendel** [comment]
I'm glad there is a proof of concept to help clarify what this looks like.

I do not like the fact that _nothing_ can ever be "just a view" with these semantics, including `series[::-1]`, `frame[col]`, `frame[:]`.  Users reasonably expect numpy semantics for these.

We should revisit the alternative "clear/simple rules" approach that is "indexing on columns always gives a view" (#33597).  This is simpler to explain/grok, simpler to implement, and not dependent on BlockManager vs ArrayManager.

**Comment by matthew-brett** [comment]
Please forgive my lack of understanding, but how does the Error-on-Write proposal here differ from the current behavior after:

```python
pd.set_option('mode.chained_assignment', 'raise')
```
?  I mean, apart from the proposal to add the extra `copy_if_needed` and `as_mutable_view` methods?

**Comment by matthew-brett** [comment]
And - rreflecting on my fairly heavy use of Numpy for about 15 years - I rarely depend on having a view onto the array, outside the direct indexing caes like `a[:10] = 1`.   I'm sure I would be happy to have the option to do that in Pandas, and sometimes using `as_mutable_view`, but, speaking as a heavy Numpy user, Copy-on-Write in Pandas would make my teaching and working life a lot easier.

**Comment by jorisvandenbossche** [comment]
@matthew-brett thanks for the feedback!

>  how does the Error-on-Write proposal here differ from the current behavior after: `pd.set_option('mode.chained_assignment', 'raise')` ?

For writing to subsets, it might not differ that much in general (there are some cases where we currently don't warn when writing to a view or copy, but that could be changed). But for the current mode to raise on chained assignment, it doesn't work the other way around. Say you have a DataFrame `df` and create a subset of it `subset = df[:5]`. Modifying `subset` will raise a warning (or error with the option), but you can currently happily modify `df` which could be reflected in `subset` without getting a warning/error. 
With the Copy-on-Write proposal, the idea is to ensure that both objects don't modify each other in both directions (both parent -> child, as child -> parent).

**Comment by matthew-brett** [comment]
@jorisvandenbossche - thanks for the clarification.  I was asking because I have used and taught `pd.set_option('mode.chained_assignment', 'raise')` - but I found that it was not uncommon for my students to run into errors that took some significant investigation to fix.   I bet you would get a fairly constant level of question and complaint with that option.

**Comment by matthew-brett** [comment]
I guess one major worry here would be that, with Copy-on-Write, Pandas 2.0 will silently return different answers from Pandas < 2.0.

Is there some way to harvest notebooks from Github, to see how often that would happen, in practice?  And in what proportion of cases the use of the < 2.0  behavior was actually a bug, where the user had not realized that:

* The assignment was going to a copy, and thus being silently discarded, or
* The assignment was going to a view, and unexpectedly modifying another series or data frame.

**Comment by matthew-brett** [comment]
Can I ask - what would it take - to get this one over the line?  And how can I help?   For example, if I can find some time, maybe funded, to do the Github notebook exploration above, and write it up, how useful would that be?  If not very useful, is there something else I can do?

**Comment by jorisvandenbossche** [comment]
@justmarkham @chendaniely Thanks a *lot* for the feedback, that's really valuable. 

I am personally happy to hear that you both think the current proposal would be more intuitive for (novice) users than the current situation (since this was one of my goals with the proposal). I also think that we should not assume our users to know (numpy) concepts of copy vs view in indexing.

> [@justmarkham] If I understood correctly, chained assignment would stop working under this proposal. I would be in favor of chained assignment no longer being allowed, even for cases in which it currently works, since you can always just use loc instead (right?).  
> [@chendaniely] I updated the above gist for what "chained indexing" looks like from the R world

So chained assignment actually _works_ in R! 
With my current proposal, chained assignment would indeed never work. 

There is one aspect related to this brought up by @shoyer on the mailing list: we could make one exception on the copy-on-write rule, i.e. accessing a single column of a DataFrame as a series would still be a view, any other indexing operation would follow copy-on-write as proposed. I recently commented about this on the mailing list at https://mail.python.org/pipermail/pandas-dev/2021-October/001395.html

This would ensure that one type of chained assignment (and possibly the most common one) keeps working: `df[col][...] = ...`. This would make the transition smoother. 
But at the same time, exactly because it is only one type of chained assignment that would work, the user needs to understand that selecting the column first works, but something like `df[row_selection][col] = ...` does not work. Although it seems a simple exception to the general rule, my feeling is that it will give rise to several potentially confusing corner cases. 

> [@chendaniely] I'm do not know about any of the performance implications of this change. And what use cases people have implemented that rely on the copy/view behaviors in pandas.

Yes, that's a good point and something I should explore in more detail (and add a section about in the document). I think in general for the majority of workflows, the performance will either stay the same or improve (because of avoiding unnecessary copies). 

There will be specific cases that might be impacted though:

* The relatively frequent case of mutating a dataframe that has a reference to another object: this can potentially become slower (if it's a case where we don't copy right now), or see a "delayed" impact (overall performance might be the same but the copy is now delayed at the moment when actually mutating)
* A potential corner case that is heavily impacted is where you mutate a dataframe repeatedly in a for loop, and where you do some operation in the for loop that triggers the copy-on-write mechanism repeatedly (I should try to come up with some example code, but can't think of a sensible example right now. For example, @shoyer showed one example in an old discussion at https://github.com/pandas-dev/pandas/pull/10973#issuecomment-137263235, but that would actually not trigger copy-on-write in my proof-of-concept implementation since I keep track of views per column instead of per full dataframe).

**Comment by nickeubank** [comment]
A little cross promotion: https://youtu.be/aBeEN2klZQE

**Comment by mattharrison** [comment]
This can't come soon enough. 

Happy to give a prototype a run through on some of my Pandas material. However, I never hit our friend `SettingWithCopyWarning` with the code I write. (I am interested in seeing speed/memory improvements though.)

As far as some code breaking, IMO you should release the next minor version before this gets pushed updating the deprecation warning to state that certain behavior will be removed, and then push a minor (major?) version with the fix.

Again, this can't come soon enough for those who haven't embraced chaining.

**Comment by nickeubank** [comment]
@jorisvandenbossche Out of curiosity, while I recognize that moving to copy on view is technically a breaking change, are you aware of any real-world examples where people are relying on pandas returning views in their code?

The rules around when pandas returns views are _so_ idiosyncratic and sensitive to apparently-unrelated code changes that I've never felt remotely safe using views, and am curious if anyone else ever does...

**Comment by matthew-brett** [comment]
I just wanted to add a huge - yes please - how can I help - when can we have this!  And @nickeubank - I agree - I never feel safe deliberately using views.   I guess the problem is that a lot of code doesn't deliberately use them, but uses them by accident.   I did offer to do a survey to find out if that happened in real life, in https://github.com/pandas-dev/pandas/issues/36195#issuecomment-905944167 - but I would in general love to know what it would take to get this in.

**Comment by jorisvandenbossche** [comment]
> Out of curiosity, while I recognize that moving to copy on view is technically a breaking change, are you aware of any real-world examples where people are relying on pandas returning views in their code?

I think that any case of _chained assignment_ that currently works (intentionally, or somewhat by accident) is an example that for sure people _do_ rely on. 

Apart from that, I fully agree that it's probably not common that people create an actual subset dataframe, and mutate that with the _intention_ to also mutate the original dataframe (which is for me a good reason to not do this, because this is not what people expect, in a large majority of cases also not what people want).

And to be clear, the chained assignment comprises both the typical case of doing something like `df["C"][df["A"] > 1] = 10` (example from my [slides](https://jorisvandenbossche.github.io/talks/2022_PyDataBerlin_pandas-copy-view/#24)), as the non-typical cases like `df[1:3]["C"] = 10`. 
But I think the first one (i.e. selecting a column, and then directly mutating that column in a chained setitem) is certainly a common pattern (although we will typically recommend to use `.loc` instead), and the fact that this will no longer work will certainly affect quite some people. (I do think that we can (first warn and then) error for this, to make this very explicit that it doesn't work, and to avoid silent failures because of this change)

**Comment by nickeubank** [comment]
> But I think the first one (i.e. selecting a column, and then directly mutating that column in a chained setitem) is certainly a common pattern (although we will typically recommend to use .loc instead), and the fact that this will no longer work will certainly affect quite some people

Yeah, ok, I can see that! The compactness of the pattern means all the weird ways a view can "break" don't come into play, and so I can see how it would be frequently used naively. Thanks!

**Comment by jorisvandenbossche** [comment]
About the "what would it take to get this in / how can I help":

First, it would actually be good to get some more explicit support from other pandas core devs (it's hard to estimate to what extent there is consensus around the latest proposal).

But apart from that:

- We need an implementation:
  - https://github.com/pandas-dev/pandas/pull/46958 implements the core of the proposal, and mainly needs people with time to review this.
  - Apart from that core implementation, there will be other work needed to fully implement the proposal. Some additional work items:
     - Implement the new behaviour for constructors (eg constructing a DataFrame/Series from an existing DataFrame/Series should follow the same rules as indexing -> behaves as a copy through CoW) 
    -  Use the lazy copy (with CoW) mechanism in more methods where it can be used (currently I only explicitly did reset_index and rename as POC)
    - Explore / update the APIs that return numpy arrays (`.values`, `to_numpy()`). Potential idea is to make the returned array read-only by default
    - ...

    (I should create a tracker issues with a task list for those items)
  - For now I only implemented the new behaviour (toggled with an option), but in addition we will also need to implement all the deprecation warnings for cases that will change in the current behaviour (initially also behind an option).
- Once the (core) implementation lands in master / a release, then it will be super useful to have people give this a try, run their code or test suites with it, etc, so we can iron out bugs / missing warnings / or discover unexpected consequences that need to be addressed/discussed.

So on the short term helping with with reviewing or some extra implementation tasks is most needed, but certainly also not super approachable to help with. 
Once there is a basic implementation it will be easier to help with testing / giving feedback (although it's certainly already welcome to give the PR branch a try as well).

**Comment by mroeschke** [comment]
I think this approached has been solidified with https://github.com/pandas-dev/pandas/pull/51463 and any future discussion should occur in https://github.com/pandas-dev/pandas/issues/48998 or in individual issues so closing
